vortex-datafusion/src/persistent/opener.rs

```diff
@@ -0,0 +1,61 @@
+use std::sync::Arc;
+
+use arrow_array::cast::as_struct_array;
+use arrow_array::RecordBatch;
+use datafusion::datasource::physical_plan::{FileMeta, FileOpenFuture, FileOpener};
+use datafusion_common::Result as DFResult;
+use datafusion_physical_expr::PhysicalExpr;
+use futures::{FutureExt as _, TryStreamExt};
+use object_store::ObjectStore;
+use vortex::IntoCanonical;
+use vortex_serde::file::reader::projections::Projection;
+use vortex_serde::file::reader::VortexBatchReaderBuilder;
+use vortex_serde::io::ObjectStoreReadAt;
+
+pub struct VortexFileOpener {
+    pub object_store: Arc<dyn ObjectStore>,
+    pub batch_size: Option<usize>,
+    pub projection: Option<Vec<usize>>,
+    pub predicate: Option<Arc<dyn PhysicalExpr>>,
+}
+
+impl FileOpener for VortexFileOpener {
+    fn open(&self, file_meta: FileMeta) -> DFResult<FileOpenFuture> {
+        let read_at =
+            ObjectStoreReadAt::new(self.object_store.clone(), file_meta.location().clone());
+
+        let mut builder = VortexBatchReaderBuilder::new(read_at);
+
+        if let Some(batch_size) = self.batch_size {
+            builder = builder.with_batch_size(batch_size);
+        }
+
+        if let Some(_predicate) = self.predicate.as_ref() {
+            log::warn!("Missing logic to turn a physical expression into a RowFilter");
+        }
+
+        if let Some(projection) = self.projection.as_ref() {
+            builder = builder.with_projection(Projection::new(projection))
+        }
+
+        DFResult::Ok(
```

Comment:
We should figure out fmt/clippy rule for this

Reply:
There's a rule, however, it's very fiddly to configure. Will have a look at some point if there's a simple setup